components/users/api/src/main/java/org/phenotips/groups/internal/UsersAndGroups.java

```diff
@@ -0,0 +1,307 @@
+            if (attachmentRefs.size() > 0) {
+                avatarURL = this.bridge.getAttachmentURL(attachmentRefs.get(0), true);
+            } else {
+                XWikiContext xcontext = this.xcontextProvider.get();
+                XWiki xwiki = xcontext.getWiki();
+                avatarURL = xwiki.getSkinFile("icons/xwiki/noavatar.png", xcontext);
+            }
+            String name = user.getName();
+            if (StringUtils.isBlank(name)) {
+                name = user.getUsername();
+            }
+            JSONObject o = createObject(user.getProfileDocument().toString(), name, avatarURL, USER);
+            resultArray.put(o);
+            if (resultArray.length() == maxResults) {
+                break;
+            }
+        }
+    }
+
+    private void runGroupsQuery(JSONArray resultArray, String queryString, String formattedInput, int maxResults)
+        throws Exception
+    {
+        List<String> queryResult = runQuery(queryString, formattedInput, maxResults);
+        for (String groupName : queryResult)
+        {
+            Group group = this.groupManager.getGroup(groupName);
+            List<AttachmentReference> attachmentRefs = this.bridge.getAttachmentReferences(group.getReference());
+            String avatarURL = "";
+            if (attachmentRefs.size() > 0) {
+                avatarURL = this.bridge.getAttachmentURL(attachmentRefs.get(0), true);
+            } else {
+                XWikiContext xcontext = this.xcontextProvider.get();
+                XWiki xwiki = xcontext.getWiki();
+                avatarURL = xwiki.getSkinFile("icons/xwiki/noavatargroup.png", xcontext);
+            }
+            JSONObject o = createObject(groupName, group.getReference().getName(), avatarURL, GROUP);
+            resultArray.put(o);
+        }
+    }
+
+    private List<String> runQuery(String queryString, String formattedInput, int resultsLimit)
+    {
+        Query query = null;
+        List<String> queryResults = null;
+        try {
+            query = this.qm.createQuery(queryString, Query.XWQL);
+            if (resultsLimit > 0) {
+                query.setLimit(resultsLimit);
+                query.setOffset(0);
+            }
+            if (StringUtils.isNotBlank(formattedInput)) {
+                query.bindValue(UsersAndGroups.INPUT_PARAMETER, formattedInput);
+            }
+            queryResults = query.execute();
+        } catch (QueryException e) {
+            this.logger.error("Error while performing query: [{}] ", queryString, e.getMessage());
+            return Collections.emptyList();
+        }
+        return queryResults;
+    }
+
+    private JSONObject createObject(String id, String value, String avatar, String type)
+    {
+        JSONObject o = new JSONObject();
+
+        StringBuilder idWithType = new StringBuilder();
+        idWithType.append(id).append(";").append(type);
+        o.put(UsersAndGroups.ID_KEY, idWithType.toString());
```

Comment:
No idea, I didn't get this far with the review. Now that you mention it, it does look wrong. I assume it has something to do with how the projects&templates code is using this service (reminder that this is where @veronikaslc extracted the code from).